src/Compression/CompressionCodecDeflate.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+#include <map>
+#include <vector>
+#include <Compression/ICompressionCodec.h>
+#include <IO/BufferWithOwnMemory.h>
+#include <IO/WriteBuffer.h>
+#include <Parsers/StringRange.h>
+#include <qpl/qpl.h>
+#include <x86intrin.h>
+namespace Poco
+{
+class Logger;
+}
+
+namespace DB
+{
+
+class DeflateJobHWPool
+{
+public:
+    DeflateJobHWPool();
+    ~DeflateJobHWPool();
+    static DeflateJobHWPool & instance();
+    static constexpr auto jobPoolSize = 1024;
+    static constexpr qpl_path_t PATH = qpl_path_hardware;
+    static qpl_job * jobPool[jobPoolSize];
+    static std::atomic_bool jobLock[jobPoolSize];
+    bool jobPoolEnabled;
+
+    bool ALWAYS_INLINE jobPoolReady()
+    {
+        return jobPoolEnabled;
+    }
+    qpl_job * ALWAYS_INLINE acquireJob(uint32_t * job_id)
+    {
+        if (jobPoolEnabled)
+        {
+            uint32_t retry = 0;
+            auto index = random(jobPoolSize);
+            while (tryLockJob(index) == false)
+            {
+                index = random(jobPoolSize);
+                retry++;
+                if (retry > jobPoolSize)
+                {
+                    return nullptr;
+                }
+            }
+            *job_id = jobPoolSize - index;
+            return jobPool[index];
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+    qpl_job * ALWAYS_INLINE releaseJob(uint32_t job_id)
+    {
+        if (jobPoolEnabled)
+        {
+            uint32_t index = jobPoolSize - job_id;
+            ReleaseJobObjectGuard _(index);
+            return jobPool[index];
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+    qpl_job * ALWAYS_INLINE getJobPtr(uint32_t job_id)
+    {
+        if (jobPoolEnabled)
+        {
+            uint32_t index = jobPoolSize - job_id;
+            return jobPool[index];
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+
+private:
+    size_t ALWAYS_INLINE random(uint32_t pool_size)
+    {
+        size_t tsc = 0;
```

Comment:
Can't we use a random generator from the C++ standard library? Performance isn't of utmost concern here and the code would be easier to read/maintain.

Reply:
Done